src/endpoint.ts

```diff
@@ -169,24 +172,59 @@ export class Endpoint<
     return this.outputSchema;
   }
 
-  public override getPositiveResponseSchema(): POS["schema"] {
-    return this.resultHandler.getPositiveResponse(this.outputSchema).schema;
+  public override getPositiveResponseSchema(): POS {
+    const apiResponse = this.resultHandler.getPositiveResponse(
+      this.outputSchema
+    );
+    return apiResponse instanceof z.ZodType ? apiResponse : apiResponse.schema;
   }
 
-  public override getNegativeResponseSchema(): NEG["schema"] {
-    return this.resultHandler.getNegativeResponse().schema;
+  public override getNegativeResponseSchema(): NEG {
+    const apiResponse = this.resultHandler.getNegativeResponse();
+    return apiResponse instanceof z.ZodType ? apiResponse : apiResponse.schema;
   }
 
   public override getInputMimeTypes() {
     return this.mimeTypes;
   }
 
+  #getMimeTypesFromApiResponse<S extends z.ZodType>(
+    subject: S | ApiResponse<S>,
+    fallback = [mimeJson]
+  ) {
+    if (subject instanceof z.ZodType) {
+      return fallback;
+    }
+    const { mimeTypes, mimeType } = subject;
+    return mimeType ? [mimeType] : mimeTypes || fallback;
+  }
+
   public override getPositiveMimeTypes() {
-    return this.resultHandler.getPositiveResponse(this.outputSchema).mimeTypes;
+    return this.#getMimeTypesFromApiResponse(
+      this.resultHandler.getPositiveResponse(this.outputSchema)
+    );
   }
```

Comment:
perhaps builder can do this.
it already supplies input mime types.
so we can cache it.

Reply:
I'm going to do the refactoring in v9